storage/mysql/log_storage.go

```diff
@@ -559,6 +553,41 @@ func (t *logTX) getActiveLogIDsInternal(sql string) ([]int64, error) {
 	return logIDs, nil
 }
 
+func (t *logTX) getLeavesByHashInternal(leafHashes [][]byte, tmpl *sql.Stmt, desc string) ([]trillian.LogLeaf, error) {
+	stx := t.tx.Stmt(tmpl)
+	var args []interface{}
+	for _, hash := range leafHashes {
+		args = append(args, interface{}([]byte(hash)))
+	}
+	args = append(args, interface{}(t.ls.logID))
+	rows, err := stx.Query(args...)
+	if err != nil {
+		glog.Warningf("Query() %s hash = %v", desc, err)
+		return nil, err
+	}
+
+	// The tree could include duplicates so we don't know how many results will be returned
+	var ret []trillian.LogLeaf
+
+	defer rows.Close()
+	for rows.Next() {
+		leaf := trillian.LogLeaf{}
+
+		if err := rows.Scan(&leaf.MerkleLeafHash, &leaf.LeafValueHash, &leaf.LeafValue, &leaf.LeafIndex); err != nil {
+			glog.Warningf("Scan() %s = %s", desc, err)
+			return nil, err
+		}
+
+		if got, want := len(leaf.MerkleLeafHash), t.ls.hashSizeBytes; got != want {
+			return nil, fmt.Errorf("Scanned leaf %s does not have hash length %d, got %d", desc, want, got)
```

Comment:
We have some ideas on how to detect corruption but we haven't implemented them yet really. It's been thought about.